docs/oppen_algorithm.md

```diff
@@ -0,0 +1,179 @@
+# Derek C. Oppen's pretty pritning algorithm
+
+Derek C. Oppen's algorithm provides an efficient solution for formatting and printing code with clean indentation that respects a specified width.
+When working with large blocks of code, ensuring proper indentation and alignment is crucial for readability and maintainability. However, achieving this while adhering to a specific line width can be a challenging task.
+This algorithm is great if you are working with code for documentation, output logging, or any situation that requires code formatting.
+
+## The printing tokens
+
+|   Name    | Description
+|-----------|------------
+| String    | String literal to be displayed.
+| Begin     | Opens a group.
+| End       | Closes a group.
+| Break     | Potential line break (adds a new line if the text doesn't fit on the line).
+| LineBreak | Guaranteed line break.
+
+### Groups
+
+A group represents a sequence of items that shall be printed together.
+A consistent group propagates a new line to all the `BREAK` tokens in the group,
+while an inconsistent group does not propagate the new line.
+
+## The algorithm
+
+The algorithm can be divided into two parts: the `scanning` and the `printing`. As their name implies, the former adds new tokens into our buffers/stacks (expanding to the right) while the latter pushes onto our `print stack` the tokens from the bottom of our buffers (shrinking from the left).
+We scan until the current length of the line is greater or equal than the maximum desired width, then we will keep on pushing onto the `print stack` the tokens that we pop from our `tokens queue` while their correspondong `size buffer` value is greater than 0.
+
+### Scan stack
+
+The `scan stack` saves the locations of all the `BEGIN`, `END`, and `BREAK` tokens, allowing the algorithm to quickly access the last `BREAK` tokens and the groups they belong to.
+
+### Print stack
+The `print stack` handles the display of the tokens, taking into account the current indentation and the available space left on the line.
+It saves only the `BEGIN` tokens, the other tokens are only printed to the output.
+
+### Tokens queue
+
+The `tokens queue`'s saves the scanned tokens that are not pushed yet to the `print stack`.
+
```

Comment:
This is a weird way to formulate it. Why aren't they pushed yet to the print stack? Does it contain things from the scan stack in the `printing phase` ?